pkg/controllers/pod/pod_controller.go

```diff
@@ -0,0 +1,310 @@
+
+		filteredPods := make(map[string]interface{})
+
+		// Get all workloadEndpoints for kubernetes orchestrator from ETCD datastore
+
+		workloadEndpoints, err := calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{
+			Orchestrator: "k8s",
+		})
+		if err != nil {
+			return filteredPods, err
+		}
+
+		// Filter out only objects that are written by policy controller
+		for _, endpoint := range workloadEndpoints.Items {
+
+			key := podConverter.GetKey(endpoint)
+			filteredPods[key] = endpoint.Metadata.Labels
+		}
+		log.Debugf("Found %d pods in calico ETCD:", len(filteredPods))
+		return filteredPods, nil
+	}
+
+
+	cacheArgs := calicocache.ResourceCacheArgs{
+		ListFunc:   listFunc,
+		Client:     calicoClient,
+		ObjectType: reflect.TypeOf(map[string]string{}), // Restrict cache to store pod labels only.
+	}
+
+	labelCache := calicocache.NewResourceCache(cacheArgs)
+	endpointCache := make(map[string]api.WorkloadEndpoint)
+
+	
+
+	// create the watcher
+	listWatcher := cache.NewListWatchFromClient(k8sClientset.Core().RESTClient(), "pods", "", fields.Everything())
+
+	// Bind the calico cache to kubernetes cache with the help of an informer. This way we make sure that
+	// whenever the kubernetes cache is updated, changes get reflected in calico cache as well.
+	indexer, informer := cache.NewIndexerInformer(listWatcher, &v1.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Infof("Got ADD event for pod: %s\n", key)
+
+			endpoint, conversionErr := podConverter.Convert(obj)
+			if conversionErr != nil {
+				log.Errorf("Error while converting %#v to endpoint.", obj)
+				return
+			}
+
+			workloadEndpoint := endpoint.(api.WorkloadEndpoint)
+			calicoKey := podConverter.GetKey(workloadEndpoint)
+			// Add workloadID:Labels in labelCache
+			labelCache.Prime(calicoKey, workloadEndpoint.Metadata.Labels)
+		},
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+
+			key, err := cache.MetaNamespaceKeyFunc(newObj)
+
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Infof("Got UPDATE event for pod: %s\n", key)
+			log.Debugf("Old object: %#v\n", oldObj)
+			log.Debugf("New object: %#v\n", newObj)
+
+			endpoint, conversionErr := podConverter.Convert(newObj)
+			if conversionErr != nil {
+				log.Errorf("Error while converting %#v to endpoint.", newObj)
+				return
+			}
+
+			workloadEndpoint := endpoint.(api.WorkloadEndpoint)
+			calicoKey := podConverter.GetKey(workloadEndpoint)
+			// Add workloadID:Labels in labelCache
+			labelCache.Set(calicoKey, workloadEndpoint.Metadata.Labels)
+		},
+		DeleteFunc: func(obj interface{}) {
+			// IndexerInformer uses a delta queue, therefore for deletes we have to use this
+			// key function.
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			log.Infof("Got DELETE event for pod: %s\n", key)
```

Comment:
This is probably too verbose - I think Add/Update/Delete logs need to be debug level.